pallets/automation-time/src/migrations.rs

```diff
@@ -320,3 +322,168 @@
 // 		}
 // 	}
 // }
+
+pub mod v4 {
+	use super::*;
+	use crate::{AccountTaskId, ScheduledTasks};
+	use frame_support::{migration::storage_key_iter, BoundedVec, Twox64Concat};
+
+	pub const OLD_STORAGE_PREFIX: &[u8] = b"ScheduledTasksV2";
+	pub const NEW_STORAGE_PREFIX: &[u8] = b"ScheduledTasksV3";
+
+	// Use weight to determine when scheduled task slot is full
+	pub struct MigrateToV4<T>(PhantomData<T>);
+	impl<T: Config> OnRuntimeUpgrade for MigrateToV4<T> {
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<(), &'static str> {
+			use frame_support::traits::OnRuntimeUpgradeHelpersExt;
+
+			// Get count per scheduled time
+			let mut pre_scheduled_count: Vec<(UnixTime, u32)> = vec![];
+			storage_key_iter::<u64, BoundedVec<AccountTaskId<T>, T::MaxTasksPerSlot>, Twox64Concat>(
+				PALLET_PREFIX,
+				OLD_STORAGE_PREFIX,
+			)
+			.for_each(|(time, task_ids)| {
+				pre_scheduled_count.push((time, task_ids.len() as u32));
+			});
+			Self::set_temp_storage::<Vec<(UnixTime, u32)>>(
+				pre_scheduled_count,
+				"pre_scheduled_count",
+			);
+
+			log::info!(
+				target: "automation-time",
+				"migration: AutomationTime storage version v4 PRE migration checks succesful!"
+			);
+
+			Ok(())
+		}
+
+		fn on_runtime_upgrade() -> Weight {
+			log::info!(target: "automation-time", "Migrating automation-time v4");
+
+			// Move all tasks from ScheduledTasksV2 to ScheduledTasksV3
+			let mut task_count = 0u64;
+			let migrated_keys_count = storage_key_iter::<UnixTime, BoundedVec<AccountTaskId<T>, T::MaxTasksPerSlot>, Twox64Concat>(
+				PALLET_PREFIX,
+				OLD_STORAGE_PREFIX,
+			)
+			.drain()
+			.map(|(time, task_ids)| {
+				let weight = task_ids.clone()
+					.into_iter()
+					.filter_map(|(account_id, task_id)| {
```

Comment:
could also use[ .fold](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fold) for this